scvi/train/_trainingplans.py

```diff
@@ -853,3 +865,215 @@ def configure_optimizers(self):
+            elbo = jnp.mean(loss_recorder.reconstruction_loss + loss_recorder.kl_local)
+            return loss, (elbo, new_model_state)
+
+        (loss, (elbo, new_model_state)), grads = jax.value_and_grad(
+            loss_fn, has_aux=True
+        )(state.params)
+        new_state = state.apply_gradients(
+            grads=grads, batch_stats=new_model_state["batch_stats"]
+        )
+        return new_state, loss, elbo, rngs
+
+    def training_step(self, batch, batch_idx):
+        if "kl_weight" in self.loss_kwargs:
+            self.loss_kwargs.update({"kl_weight": self.kl_weight})
+        self.state, loss, elbo, self.rngs = self.jit_training_step(
+            self.state,
+            batch,
+            self.rngs,
+            loss_kwargs=self.loss_kwargs,
+        )
+        loss = torch.tensor(jax.device_get(loss))
+        elbo = torch.tensor(jax.device_get(elbo))
+        # TODO: Better way to get batch size
+        self.log(
+            "train_loss",
+            loss,
+            on_epoch=True,
+            batch_size=batch[REGISTRY_KEYS.X_KEY].shape[0],
+        )
+        self.log(
+            "elbo_train",
+            elbo,
+            on_epoch=True,
+            batch_size=batch[REGISTRY_KEYS.X_KEY].shape[0],
+        )
+
+    @partial(jax.jit, static_argnums=(0,))
+    def jit_validation_step(
+        self,
+        state: TrainState,
+        batch: Dict[str, jnp.ndarray],
+        rngs: Dict[str, jnp.ndarray],
+        **kwargs,
+    ):
+        module = self.validation_module
+        rngs = {k: random.split(v)[1] for k, v in rngs.items()}
+        vars_in = {"params": state.params, "batch_stats": state.batch_stats}
+        outputs = module.apply(vars_in, batch, rngs=rngs, **kwargs)
+        loss_recorder = outputs[2]
+        loss = loss_recorder.loss
+        elbo = jnp.mean(loss_recorder.reconstruction_loss + loss_recorder.kl_local)
+
+        return loss, elbo
+
+    def validation_step(self, batch, batch_idx):
+        loss, elbo = self.jit_validation_step(
+            self.state, batch, self.rngs, loss_kwargs=self.loss_kwargs
+        )
+        loss = torch.tensor(jax.device_get(loss))
+        elbo = torch.tensor(jax.device_get(elbo))
+        self.log(
+            "validation_loss",
+            loss,
+            on_epoch=True,
+            batch_size=batch[REGISTRY_KEYS.X_KEY].shape[0],
+        )
+        self.log(
+            "elbo_validation",
+            elbo,
+            on_epoch=True,
+            batch_size=batch[REGISTRY_KEYS.X_KEY].shape[0],
+        )
+
+    @property
+    def kl_weight(self):
+        """Scaling factor on KL divergence during training."""
+        return _compute_kl_weight(
+            self.current_epoch,
+            self.global_step,
+            self.n_epochs_kl_warmup,
+            self.n_steps_kl_warmup,
+        )
+
+    def transfer_batch_to_device(self, batch, device, dataloader_idx):
+        """Bypass Pytorch Lightning device management."""
```

Comment:
where is the jax device management? Is it determined by the device of the PRNG key?

Reply:
issue created for this https://github.com/scverse/scvi-tools/issues/1553